### DISCUSSION TITLE
Additional property simplification

### DISCUSSION THREAD

**Discussion by KingTempest07** [dotnet/csharplang#7409] [Language Ideas] [answered]
This is *mainly* an idea for simplifying property creation when the property is only acting as a redirection to another value. This is shorthand and is not necessary, but I think it would be nice to have and fairly easy to implement (sort of like implicit `new()` expressions).

Additionally, there are two separate ideas and one extension idea, but they are closely related, so I put them all here.

---

When creating an interface member, they are always forced to be properties. Because of this, you sometimes might implement the member and have it only act as a redirection to another field in your implementing class, like this:

```
public Type[] types;
Type[] IFoo._types {
    get => types;
    set => types = value;
}
```

To me, the `= value` part is redundant, assuming it's only going to act as a middle-man but not change anything. 

So the first idea I'm proposing is that it could be simplified down to this:

```
public Type[] types;
Type[] IFoo._types {
    get => types;
    set => types;
}
```

If the user ever wants it to change anything, they can always add the equals operator back and still utilize the value parameter.

Essentially, I want to implicitly change the value of a variable directly following `set =>` without needing to say `= value`

This idea would also apply even if you did want to change the get value, as shown here:

```
public float number;
float IFoo.bar {
    get => number * 10f;
    set => number;
}
```

---

Additionally, if you were to go even further, using a new keyword (I'm going to use `use` for this example), you could simplify it to one line that essentially says the same thing as the first two examples:

```
public Type[] types;
Type[] IFoo._types {
    use types;
}
```

Or put more simply, maybe even something inline like what getters have (although I don't like using the colon with this particular case):

```
public Type[] types;
Type[] IFoo._types : types;
```

Essentially, this just makes it shorthand for exactly what the first two examples already do. This means the underlying code would just be this:

```
public Type[] types;
Type[] IFoo._types {
    get => types;
    set => types = value;
}
```

If you want to change individual accessibility modifiers or modify the set values in some other way, then you should not be using shorthand and instead should use normal, written-out getters and setters.

My two name ideas for the keyword are `use` or `redirect`, but `use` could be conflated with `using`, so `redirect` is my preference. But if this becomes a feature, there might be better or more accurate names (or the keyword could just be skipped altogether in favor of an operator).

---

Also, note that I said you should not use this new keyword to modify the *set* values in the previous part. This is because I have another, less likely-to-get-implemented idea for a scenario you could use this. This is a little harder to read, and just a thought I had on the fly, but here it is:

```
public float foo = 5f;
public float bar {
    use foo * 10f;
}

// using an inline operator:
// public float bar : foo * 10f;
```

Here's what the underlying code would look like:

```
public float foo = 5f;
public float bar {
    get => foo * 10f;
    set => foo = value;
}
```

Essentially: 

- When *getting* `bar`, it will return the value of `foo` multiplied by 10 (50)
- When *setting* `bar`, it will still directly assign the value to `foo` (without multiplying it by 10)
  - Ex.: after running `bar = 10f`, `foo` will be set to 10, and `bar` will still return `foo * 10f` (which now equals 100)

It's just more shorthand for getters and setters, but I still think it would be nice to have. On top of that, there are probably a lot more applications for this functionality that I haven't thought about yet.

---

My reasoning for wanting the first two ideas, however, is pretty simple, but it very well could just be due to me using Unity (which uses an older version of C#, so I won't benefit from a change for a while) and may not warrant a change.

Essentially, when implementing an interface member, you need to have two separate values in Unity to have one show up in the inspector. This is because Unity cannot render properties, so instead, you need to make the property get and set the value of a serialized field (which will appear in the inspector so that its values can be changed). 

Typing out the same getters and setters has gotten annoying, and it has started unnecessarily bloating my code, making it harder to read at a glance. That's exactly what shorthand is meant to fix, and that's what I hope my ideas could do.

I put this here, though, in case this is a common problem with *regular* C# as well that anyone might benefit from having a cleaner and easier way of doing it. That and the fact that eventually, Unity will update to the latest version, and that feature would be very nice to have when they do.

I'm okay with none of these ideas getting implemented, but they would be welcome changes for me. But if anyone has any ideas to make this concept better or more useful in *normal* C#, feel free to add your ideas.

**Comment by huoyaoyuan** [comment]
How about #140?

**Reply by KingTempest07** [reply]
From what I can tell, this is different from what I would want. 

In Unity, the field needs to be a separate, serializable member. 

That seems to just allow you to access the automatically-created field from inside the property (while keeping it in local scope) unless I'm understanding it incorrectly. 

I think so far, anyway, that the consensus on my idea is that it's unnecessary (seeing as there are only downvotes right now), but I can't tell when no one says anything lol. I kind of foresaw it being an oversimplification and maybe just more work than what it's worth, and even though I still think something similar would be nice, I can live with it staying as is.

**Comment by TahirAhmadov** [comment]
There is a discussion or an issue for this, but I couldn't find it. Basically, the existing proposal calls for something like:
```cs
public int Prop <=> this._backingField; // _backingField is not readonly
public int Prop <=> this._someObj.AnotherProp; // AnotherProp is get/set
```

**Reply by KingTempest07** [reply]
Ah ok. I couldnt find anything either, but I'm not surprised that I'm not the first to think of this. 

That way is definitely way cleaner, though, and also resembles the get operator `=>` better.

**Reply by CyrusNajmabadi** [reply]
SPACESHIP 🚀

**Reply by TahirAhmadov** [reply]
> SPACESHIP 🚀

??

**Reply by CyrusNajmabadi** [reply]
https://en.m.wikipedia.org/wiki/Three-way_comparison

**Reply by TahirAhmadov** [reply]
Are you saying there are discussions about adding that operator to C#? If that were to happen, would it conflict with the property get/set shorthand syntax `<=>`?

**Reply by Joe4evr** [reply]
@TahirAhmadov I believe you're talking about #1278, in which an ex-LDM member already explains why that has a functionally zero chance of happening:
> Recently, we have been working on a number of language changes to make it easier to program with immutable data. Part of the strategy is to make new shorter shortcuts for readonly properties. If the syntax for a read-write property is shorter than the syntax for a read-only property, then we have failed, and the changes would not encourage programming with immutable data as we intended.

**Comment by DavidArno** [comment, marked as answer]
> Essentially, when implementing an interface member, you need to have two separate values in Unity to have one show up in the inspector. This is because Unity cannot render properties, so instead, you need to make the property get and set the value of a serialized field (which will appear in the inspector so that its values can be changed).

This sounds very much like there is a limitation in Unity and that you are hoping the C# language could be changed to work around that limitation. I can't find it, but I'm fairly sure that the team have clearly stated before that they will not modify C# to fix limitations of Unity. It's a Unity problem and therefore a problem that Unity needs to fix. 🤷

**Reply by KingTempest07** [reply]
Okay. I had a feeling this might just be a Unity problem, but I figured I would put this here just in case. I'll close this discussion.

Edit: this *may* not be just a Unity problem from the context of another thread. Supposedly it was an idea, but didnt happen due to the ease of writing it making immutable data less appealing. Im still going to keep this as the answer, though, seeing as no one else here seem to have this problem anyways.
